Set Default IBL for Model Element

Phipson opened the pull request:
#### a1b6a142ec161ca904cdcf8f0db1cac8e7ae0bc3
<pre>
Set Default IBL for Model Element
<a href="https://bugs.webkit.org/show_bug.cgi?id=289626">https://bugs.webkit.org/show_bug.cgi?id=289626</a>
<a href="https://rdar.apple.com/146805748">rdar://146805748</a>

Reviewed by Ada Chan.

This PR adds a fallback resource for lighting inside the portal if the developer does not supply a default environment map
for the page. With this default lighting option, models will look nicer and less dark on the page by default.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::applyEnvironmentMapDataAndRelease):i
- Adds call to apply fallback IBL in case environment map data doesn&apos;t exist

* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.applyIBL(_:completion:)):
(WKSRKEntity.applyDefaultIBL()):
- Adds API to apply a default IBL

* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:

Canonical link: <a href="https://commits.webkit.org/292051@main">https://commits.webkit.org/292051@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/246ea9cdb4ceaff1a4d5bc07a93867da2ad5cfc9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94796 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99812 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45285 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14662 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22804 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72311 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29614 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10934 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85600 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52643 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10629 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44625 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80855 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3412 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101855 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21824 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15942 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81313 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22071 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81617 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80693 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20161 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25247 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2659 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/15057 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21801 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26921 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21461 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24935 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -734,6 +734,10 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
             send(Messages::ModelProcessModelPlayer::DidFinishEnvironmentMapLoading(true));
```
We should remove `-removeIBL` from `WKSRKEntity` and just use `-applyDefaultIBL` if something is wrong with the webpage specified environment map, if the environment map is removed, or if no environment map is specified.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -734,6 +734,10 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    } else {
+        [m_modelRKEntity applyDefaultIBLWithCompletion:^(BOOL succeeded) {
+            send(Messages::ModelProcessModelPlayer::DidFinishEnvironmentMapLoading(succeeded));
+        }];
```
I don't think `send(Messages::ModelProcessModelPlayer::DidFinishEnvironmentMapLoading(succeeded));` is necessary when there's no `m_transientEnvironmentMapData`, but I guess since `HTMLModelElement::didFinishEnvironmentMapLoading()` is a no-op if `m_environmentMapURL` is empty, it doesn't hurt, just an extra IPC call.

Phipson replied:
For simplicity I'll just remove the No-op call for the time being until we find a need for it

achan00 commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -256,7 +260,24 @@ public final class WKSRKEntity: NSObject {
                 completion(true)
             } catch {
                 Logger.realityKitEntity.error("Cannot load environment resource from CGImage")
```
Nit: could update the error message to mention the default IBL is loaded instead.

achan00 commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -234,12 +234,12 @@ public final class WKSRKEntity: NSObject {
     @objc(applyIBLData:withCompletion:) public func applyIBL(data: Data, completion: @escaping (Bool) -> Void) {
         guard let imageSource = CGImageSourceCreateWithData(data as CFData, nil) else {
-            Logger.realityKitEntity.error("Cannot get CGImageSource from IBL image data")
+            Logger.realityKitEntity.error("Cannot get CGImageSource from IBL image data. Applying default IBL instead.")
```
Now that you'v moved the logic to apply default IBL on failure out of this method, then we shouldn't change the error message to mention that we are applying default IBL.

achan00 commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -234,12 +234,12 @@ public final class WKSRKEntity: NSObject {
         }
         guard let cgImage = CGImageSourceCreateImageAtIndex(imageSource, 0, nil) else {
-            Logger.realityKitEntity.error("Cannot get CGImage from CGImageSource")
+            Logger.realityKitEntity.error("Cannot get CGImage from CGImageSource. Applying default IBL instead.")
```
Ditto.

achan00 commented on Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
```diff
@@ -255,15 +255,30 @@ public final class WKSRKEntity: NSObject {
                 completion(true)
             } catch {
-                Logger.realityKitEntity.error("Cannot load environment resource from CGImage")
+                Logger.realityKitEntity.error("Cannot load environment resource from CGImage. Applying default IBL instead.")
```
Ditto.

achan00 reviewed (approved):
LGTM if you'd like to fix the log message comment and if the builds look green.

Phipson committed:
Set Default IBL for Model Element
https://bugs.webkit.org/show_bug.cgi?id=289626
rdar://146805748

Reviewed by Ada Chan.

This PR adds a fallback resource for lighting inside the portal if the developer does not supply a default environment map
for the page. With this default lighting option, models will look nicer and less dark on the page by default.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::applyEnvironmentMapDataAndRelease):i
- Adds call to apply fallback IBL in case environment map data doesn't exist

* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.applyIBL(_:completion:)):
(WKSRKEntity.applyDefaultIBL()):
- Adds API to apply a default IBL

* Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:

Canonical link: https://commits.webkit.org/292051@main
